docs/specification.rst

```diff
@@ -63,606 +61,407 @@ For example,
- - ``typing.Never`` and ``typing.NoReturn`` also called `bottom type <https://en.wikipedia.org/wiki/Bottom_type>`_
- - ``None``
-
-If such a type is used within an intersection this intersection shall evaluate to ``Never``.
-
-The reasoning behind this is that these types can't be subtyped and shouldn't be dynamically
-extended.
-Doing this early prevents issues during subtyping or assignments checks.
 
-::
+``Intersection`` does not forbid any incompatibility of type parameters
+(Neither statically or at runtime).
 
-    from typing import TypeVar, reveal_type
 
-    T = TypeVar("T")
+Runtime specification behavior
+------------------------------
 
-    class Enhanced:
-        is_great: bool
+At runtime, ``Intersection[*Ts]`` and ``Intersection[TypeOne, TypeTwo]`` each create an
+object which can be introspected consistent with the methods provided for type
+introspection in ``typing`` such as, but not limited to, ``get_origin``.
 
 
-    def enhance(cls: type[T]) -> type[T & Enhanced]:
-        class New(cls, Enhanced):
-            ...
 
-        return New
+Rationale
+=========
 
-    reveal_type(enhance(str))  # okay
-    reveal_type(enhance(None))  # raises a TypeError on runtime, should be flagged by TypeCheckers
+The intersection discussion was long and complex, with many edge cases explored
+to determine the feasibility of an intersection. In this section I will
+summarize why certain design decisions were made.
 
-It is important to note that once a type checker evaluated anything to ``Never`` within an
-intersection it can stop further evaluations an return ``Never``.
-This way a lot of edge cases by mixin types that can't be mixed are handled easily.
+Inheritance
+-----------
 
-Handling Callables
-------------------
-Every Callable within an intersection shall be treated like a ``def __call__()`` Protocol.
+This was never very controversial - the original idea was to have intersection
+reflect the way that union works. For ``Union[A,B]``, it follows that:
 
 ::
 
-    from typing import Protocol, Callable
+    x: Union[A,B]
+    isinstance(x, A) or isinstance(x, B) # Always true
 
-    MyCallable = Callable[[str, int], float]
+So for Intersection it also follows that:
 
-    class CallProto:
-        def __call__(a: str, b: int) -> float: ...
+::
 
-    # Type Checker should perform the following conversion
-    # T & MyCallable => T & CallProto
+    x: Intersection[A,B]
+    isinstance(x, A) and isinstance(x,B) # Always true
 
-This way the ``overload`` mechanism described below can be used.
+The only discrepancy here is that would mean it must be a class that inherits from
+both A and B. In inheritance the order matters, so suddenly this means that the order
+matters for intersection as well.
 
+Ordering
+--------
 
-Protocol Reduction
-------------------
+Introducing ordering has many benefits, including the fact that it simplifies and
```

Comment:
I would say `A&B` is a subtype of `B&A` if `A` and `B` differ (have no overlapping non dunder properties) or the overlapping properties are identical. 
If they have different properties they aren't identical.

If I got the ordering correct `A & Any` is kind of `Any` because `Any` overwrites everything.
But for `Any & A`  is all properties of `A` but everything that is not a property of `A` is `Any`

Reply:
Point of clarity, it needs to be treated as an unknown subtype of each of `A` and `Any`, and some things within that can produce static type errors where being a subtype of A provides enough information. This came up in prior discussion, was proven to be so, and over on discourse, when discussing subtyping of Any, ignoring the other supertypes was loudly untenable for language servers.

Reply:
Thanks for chiming in Michael. Like I said, I'm sure this has been discussed before.

Can you link me to the relevant Discourse thread to avoid repeating the argument here?

My gut feeling, still, is that intersections are best left unordered in the PEP, because the ordering-aware design is complex, unintuitive and does not match the behavior of intersection types in other mainstream languages (e.g. TS and Java/Kotlin).

Reply:
I agree on them needing to be unordered. The ordering attempts were (both my previous ordering attempt which I went back to the drawing board on, and Mark's here)  attempts to simplify the specification with an approximation of behavior that should be equivalent, it doesn't work though The discourse thread starting here https://discuss.python.org/t/take-2-rules-for-subclassing-any/47981/5 Has Eric's objection to ignoring the other bases with multiple supertypes, this with multiple inheritance involving Any, rather than intersections involving Any, but the problems are the same here.

If you don't end up reading the whole thread, this section of my post has a way to handle this that works https://discuss.python.org/t/take-2-rules-for-subclassing-any/47981/7#further-detailing-lsp-compatible-any-and-type-bounds-2 for both intersections and multiple inheritance, but I was steered away from it at first to try and get a smaller change to work, we found that the smaller change did not work.

___

I don't think TS will be a good model here (not rehashing that right now either) and we want to look at Kotlin and Elixir. In particular, Elixir has an in-progress handling of this that solves this well even with `dynamic()` by considering function domains. https://elixir-lang.org/blog/2023/09/20/strong-arrows-gradual-typing/

This gives us an answer for gradual callables/methods in an intersection:

```
(int) -> int & (Any) -> Any
```
Here, because Any is for compatibility, we can compare function domains and arrive at....
```
(int) -> int & (Any & ~int) -> Any
```

I think the best way forward is just the subtyping rules paired with an assumption that since Any is for gradual typing and not yet-expressible concepts, Any can't contain something that would be a type error if expressed, and only represents a compatible unknown.

Reply:
> I think the best way forward is just the subtyping rules paired with an assumption that since Any is for gradual typing and not yet-expressible concepts, Any can't contain something that would be a type error if expressed, and only represents a compatible unknown.

Yeah, I like this. Do we need to build consensus before refining the PEP in the way you described?

Reply:
> Do we need to build consensus before refining the PEP in the way you described?

Probably should try to, otherwise we have competing approaches being worked on simultaneously.

Reply:
I guess my main point of contention here is that the subtyping rules seem insufficient to describe all cases without a lot of further clarification. Or at least, whenever I try to apply them to some of the examples it leads to situations that don't make sense without further resolution.

> My gut feeling, still, is that intersections are best left unordered in the PEP, because the ordering-aware design is complex, unintuitive and does not match the behavior of intersection types in other mainstream languages (e.g. TS and Java/Kotlin).

The not match other languages is a fair enough comment, but I actually think in the context of python ordered intersections are easier to understand than unordered ones. Perhaps I haven't explained it well, but really by introducing ordering any edge case that might be encountered immediately has an intuitive explanation - and there are analogies to explain it.

I think really the issue here is about expectation - it might be that this specification behaves in a way that the community finds unexpected, but I guess my hope is that by explaining it clearly this can be resolved. @superbobry Did the "how to teach" section make sense?

Reply:
subtyping rules seem to work fine for me, check latest comment in #17 for how this works for callables